### ISSUE TITLE
Back jump doesn't do well in some cases

### ISSUE THREAD

**Issue by frostming** [sarugaku/resolvelib#180]
Here are the package set to be resolved

```
1. A==1
2. B  # B@2 depends on D@2 and B@1 depends on D@1
3. C
4. D  # D@2 depends on A@2 and D@1 depends on A@1
```
Assume the resolver resolves packages from top to down, when resolving D, the current pins are:

```
A@1
B@2
C@x
```

Because `B@2` is pinned which constrains `D==2`, we can't find a compatible version of `D` that is also happy with `A@1`, so conflicts happen and we back jump and pop the last pinned package `C`, which is totally unrelated to the incompatibilities.

However, due to the line in `backjump` implementation: https://github.com/sarugaku/resolvelib/blob/51dbc929523577aea960dd3672df49c9a428ae7d/src/resolvelib/resolvers/resolution.py#L327-L330

The backjump quits very early and fallbacks to backtrack. It would be extremely slow if traversing C is super expensive. But it's obvious we should jump over to the state of `B` and try other versions. My question is, does the quit condition make sense in any case I am not aware of or should we fix it?

@notatallshaw 

- - -

To be concrete, `A: h11, B: httpcore, C: Pillow, D: httpx`

**Comment by notatallshaw** [comment]
Yeah, the exit is probably too aggressive, but there are definitely cases that incorrectly reach resolution impossible without it.

I have at least a few real world examples to try against if you have an idea how to loosen this without causing incorrect resolutions.

**Comment by frostming** [comment]
> I have at least a few real world examples to try against if you have an idea how to loosen this without causing incorrect resolutions.

I think we previously proposed to save the states first, and if no state that can jump is found, we restore to backtrack.

**Comment by notatallshaw** [comment]
Btw, I have an _intuition_ that if the backtrack causes are disjoint, i.e every cause is incomparable with some other cause (or set of causes), resolvelib could aggressively backjump. But currently the provider has no way to indicate to the resolver disjointness.

I will experiment this week with disjointness and aggressively backjumping, to see if the intuition is correct.

**Comment by notatallshaw** [comment]
> I think we previously proposed to save the states first, and if no state that can jump is found, restore the states and fallback to backtrack.

Yeah, it wasn't very elegant though, for example it made complicated impossible resolutions much slower.

**Comment by notatallshaw** [comment]
My understanding is the way backjumping was originally implemented in resolvelib did not actually meet the prerequisites for backjumping to be logically sound. In particular a resolvelib "cause" is distinct from what a "cause" might be in satisfiability theory, for example you might end up with "causes" in resolvelib that have the constraints `numpy>1.3`, `numpy>2`, `numpy<2`, but I _think_ the logical causes should be reduced to the constraints `numpy>2`, `numpy<2`.

I was hoping to introduce a new API that allowed the provider to filter these causes to only the logical ones, but I realized there's another case that I wasn't considering (https://github.com/sarugaku/resolvelib/issues/171#issuecomment-2530035387) and further I'm also concerned that the "mathematics" of pre-releases will have an unknown impact on trying to implement classical satisfiability theory (e.g. https://discuss.python.org/t/proposal-intersect-and-disjoint-operations-for-python-version-specifiers/71888/1).

All this is a long winded way to say I'm moving towards the opinion of introducing a save state. My thoughts are:

1. Only create 1 save state
2. Create that save state as late as possible
3. Only revert to the save state if resolution impossible
4. Give the provider no options to simplify the logic(?)

I think we would call this "optimistic backjumping", we backjump on the assumption that resolvelib causes are _close enough_ to logical satisfiability causes, and we revert in the face of resolution impossible.

I am going to work on making a PR and running it against scenarios in https://github.com/notatallshaw/Pip-Resolution-Scenarios-and-Benchmarks to see what the performance wins are.

**Comment by notatallshaw** [comment]
I made a [branch](https://github.com/notatallshaw/pip/tree/optimistic-backjumping) that adds a commit to add optimistic backjumping on top of resolvelib 1.1 (https://github.com/pypa/pip/pull/13001) and then ran scenarios to compare them.

It comes with a big wins and a big loss, here are the statistics comparing the resolvelib 1.1 branch to the optimistic backjumping branch:

<summary>

Output from comparison using tools from [Pip-Resolution-Scenarios-and-Benchmarks](https://github.com/notatallshaw/Pip-Resolution-Scenarios-and-Benchmarks):

<details>

```
Difference for scenario scenarios/problematic.toml - autogluon:
        Both failed: Build Failure
        Distinct Wheels visisted: 162 -> 163 (100.62%)
        Total visisted packages: 2398 -> 2184 (91.08%)
        Total visisted requirements: 22286 -> 19708 (88.43%)
        Total rejected requirements: 12791 -> 11089 (86.69%)
        Total pinned packages: 729 -> 740 (101.51%)
        Total rounds: 1283 -> 1220 (95.09%)

Difference for scenario scenarios/problematic.toml - pip-issue-12754:
        Both failed: Resolution Too Deep
        Distinct Sdists visisted: 40 -> 81 (202.50%)
        Distinct Wheels visisted: 82 -> 23 (28.05%)
        Total visisted packages: 15004 -> 19031 (126.84%)
        Total visisted requirements: 57744 -> 60022 (103.94%)
        Total rejected requirements: 55920 -> 52594 (94.05%)
        Total pinned packages: 8029 -> 8551 (106.50%)
        Total rounds: 15019 -> 15012 (99.95%)

Difference for scenario scenarios/problematic.toml - TTS:
        Total visisted packages: 398 -> 383 (96.23%)
        Total visisted requirements: 1099 -> 1084 (98.64%)
        Total rejected requirements: 581 -> 533 (91.74%)
        Total rounds: 300 -> 295 (98.33%)

Difference for scenario scenarios/problematic.toml - boto3-urllib3-transient:
        Success: False -> True
        Failure Reason: Resolution Too Deep -> None
        Distinct Sdists visisted: 28 -> 0 (0.00%)
        Distinct Wheels visisted: 2604 -> 141 (5.41%)
        Total visisted packages: 419687 -> 297 (0.07%)
        Total visisted requirements: 1436990 -> 665 (0.05%)
        Total rejected requirements: 32320 -> 57 (0.18%)
        Total pinned packages: 7584 -> 277 (3.65%)
        Total rounds: 15001 -> 281 (1.87%)

Difference for scenario scenarios/problematic.toml - kedro-test:
        Total visisted packages: 3797 -> 3843 (101.21%)
        Total visisted requirements: 12093 -> 12395 (102.50%)
        Total rejected requirements: 3006 -> 3048 (101.40%)
        Total pinned packages: 2051 -> 2084 (101.61%)
        Total rounds: 2774 -> 2816 (101.51%)

Difference for scenario scenarios/problematic.toml - sphinx:
        Failure Reason: Resolution Too Deep -> Build Failure
        Distinct Sdists visisted: 1 -> 4 (400.00%)
        Distinct Wheels visisted: 124 -> 135 (108.87%)
        Total visisted packages: 97751 -> 46035 (47.09%)
        Total visisted requirements: 135827 -> 64137 (47.22%)
        Total rejected requirements: 137900 -> 64823 (47.01%)
        Total pinned packages: 8118 -> 3904 (48.09%)
        Total rounds: 15013 -> 7145 (47.59%)

Difference for scenario scenarios/problematic.toml - backtracks-to-old-scipy:
        Both failed: Build Failure
        Total visisted packages: 235 -> 211 (89.79%)
        Total visisted requirements: 967 -> 679 (70.22%)
        Total rejected requirements: 316 -> 236 (74.68%)
        Total rounds: 99 -> 91 (91.92%)

Difference for scenario scenarios/problematic.toml - apache-airflow-google-cloud:
        Success: True -> False
        Failure Reason: None -> Resolution Too Deep
        Distinct Wheels visisted: 179 -> 195 (108.94%)
        Total visisted packages: 184 -> 16791 (9125.54%)
        Total visisted requirements: 520 -> 31559 (6069.04%)
        Total rejected requirements: 36 -> 3606 (10016.67%)
        Total pinned packages: 163 -> 14853 (9112.27%)
        Total rounds: 165 -> 15007 (9095.15%)

Difference for scenario scenarios/problematic.toml - django-stubs:
        Total visisted packages: 38 -> 30 (78.95%)
        Total visisted requirements: 190 -> 135 (71.05%)
        Total rejected requirements: 21 -> 10 (47.62%)
        Total rounds: 26 -> 25 (96.15%)

Difference for scenario scenarios/problematic.toml - kauldron-docs:
        Success: False -> True
        Failure Reason: Resolution Too Deep -> None
        Distinct Sdists visisted: 17 -> 6 (35.29%)
        Distinct Wheels visisted: 323 -> 303 (93.81%)
        Total visisted packages: 119167 -> 4565 (3.83%)
        Total visisted requirements: 372168 -> 27670 (7.43%)
        Total rejected requirements: 48082 -> 1812 (3.77%)
        Total pinned packages: 8605 -> 2362 (27.45%)
        Total rounds: 15005 -> 2830 (18.86%)
```

</details>

</summary>

Some details to note, my tooling stops resolution at ~15k rounds as so far I haven't found a resolution that completes between 15k and 200k (pip's max default) rounds.

There were no resolution differences for either popular packages or big packages, all the resolution differences were related to problematic resolutions.

Generally performance was much better, in-particular [boto3-urllib3-transient](https://github.com/notatallshaw/Pip-Resolution-Scenarios-and-Benchmarks/blob/main/scenarios/problematic.toml#L58) and [kauldron-docs](https://github.com/notatallshaw/Pip-Resolution-Scenarios-and-Benchmarks/blob/main/scenarios/problematic.toml#L240) went from resolution too deep to quickly solvable which is the best win. And [sphinx](https://github.com/notatallshaw/Pip-Resolution-Scenarios-and-Benchmarks/blob/main/scenarios/problematic.toml#L130) went from resolution too deep to build failure, which I consider a win because it gives the user information to apply lower bounds.

However, [apache-airflow-google-cloud](https://github.com/notatallshaw/Pip-Resolution-Scenarios-and-Benchmarks/blob/main/scenarios/problematic.toml#L221) went from quickly completing to resolution too deep, which I consider to be the worst failure, this is the same error that the current pip 24.3 produces, and resolvelib 1.1 was going to solve this. I suspect that optimistic backjumping has skipped over an important branch in the resolution graph and even with optimistic backjumping it can not reach resolution impossible before the 15k rounds in my tooling (nor even the 200k round in pip 24.3).

I have some bug fixes and further optimizations to pip I want to test against to see if this improves the "apache-airflow-google-cloud" situation at all, but due to obligations I don't think I will have anymore time to dedicate to this in 2024. Currently this leaves me on the fence about optimistic backjumping.

**Comment by notatallshaw** [comment]
Here's my current thinking:

1. The aim should be to enable proper backjumping (without the need to fallback) because it's clearly more efficient
2. Optimistic backjumping is a step forward for many use cases, even if it's not a 100% win, it should probably be enabled

With regards to "1." I think resolvelib needs to be slightly redesigned, specifically rather than [attempting to pin one criterion](https://github.com/sarugaku/resolvelib/blob/main/src/resolvelib/resolvers/resolution.py#L205) at a time I think we would need to pass all criterion to the provider at a time and ask "which ones are unsatisfiable?" (either because they are mutually exclusive or because no version matches their range). This would be a significant change to the API and would require a resolvelib 2.0.

For "2." I have been testing optimizations on the provider side to see if we can overcome this one known example where it reaches ResolutionTooDeep, implementing something like https://github.com/astral-sh/uv/pull/9843 looks promising, I had one example where a partially implemented version of it resolved, I will keep working on this. In the meantime I will raise a PR on resolvelibs side.

**Comment by frostming** [comment]
Feel free to do 1, the whole implementation is hidden behind a private `Resolution` class

**Comment by notatallshaw** [comment]
I'm fairly sure it will require a change to the provider API to work though, I'll start taking a more serious look in a few weeks.

**Comment by notatallshaw** [comment]
I've been reading some literature and I'm a little pessimistic resolvelib can fully and safely implement backjumping while remaining generic and not having a bunch of hidden requitements the provider must guarantee. If we look at something like pubgrub, it requires keeping a history of "facts", not just the current causes, and those facts must represent requirements in some distinct canonical form.

Packaging has a [bug](https://github.com/pypa/packaging/pull/872) in it which I think would functionaly contradicts specifiers from a logical canonical form you come come up with. But even then the Python packaging specifier standard is so unusual you have to carefully work through all the assumptions of any specific resolution algorithm. My biggest concern is that for pre-releases the specifier `<V` doesn't preserve ordinality and isn't transitive, so you have to be **very** careful on how you handle it.

So, I've come back to thinking about how to implement the fallback idea I had and raised a PR for (https://github.com/sarugaku/resolvelib/pull/183).  Currently it works like so:

1. Start resolving
2. The first time an optimistic backjump occurs save the state
3. If a ResolutionImpossible occurs revert the state and continue without optimistic backjumping

However, I found a case where ResolutionTooDeep occured under optimistic backjumping but *not* regular backjumping, so my modified plan is this:

1. Start resolving
2. The first time an optimistic backjump occurs save the state, and record what round it is
3. If a ResolutionImpossible occurs revert the state and continue without optimistic backjumping
4. If the number of rounds exceeds 50% of the remaining rounds when optimistic backjumping started revert the state and continue without optimistic backjumping

So, working through an example of this proposed additional behavior:

1. Start resolving with 100 maximum rounds
2. After 20 rounds an optimistic backjump is made
3. Optimistic backjumping occurs for another 40 rounds (100 - 20)/2 and finds nothing
4. The state is reverted back to step 2 and regular backjumping is allowed to continue for another 40 rounds

This should provide a generally faster resolution, while allowing fallback in both cases that the optimistic backjumping fails to find a solution or starts taking too long even though regular backjumping rounds have found a solution quickly. We would not expose any of the details to the user or the provider, so this optimization could be removed in the future if a better solution is found, though I was thinking of storing the exact ratio of allowed optimistic backjumping as a module level constant so that an advanced user could tweak it without it become part of the API.

Assuming no objection, I will start working on this shortly.

**Comment by notatallshaw** [comment]
While max rounds becomes more important for some cases I don't think that it needs to be more, or less, carefully, or less, selected than right now.

If you select a large value for max rounds, like I assume most users do today, then some scenarios will take a long time, but under the original suggestions those scenarios would have just got ResoltionTooDeep. And many other scenarios will not hit ResolutioTooDeep, while they do today.

If you select a small value for max rounds then, like today, many scenarios might hit ResoltionTooDeep, but probably less than today due to optimistic backjumping.
